### PR TITLE
Fixing up build publishing to be more organized

### LINKED ISSUE #3120: Refactor the Publishing code to be more organized
# Overview
When we publish built code and installers today, compiled units are compressed into archives and put into the same root folder in one Azure Blob Container (ie ```<root>\<cli_version>``` or ```<root>\<other version>```). This requires consumers of the archives (or installers) to understand the product difference between versions 1.0.0-rc2-3002520 and 1.0.0-rc2-002520.

# Proposal
In order to organize the intermediate and final build outputs (archive files and installers, respectively), I propose the following:

1. Abandon the ```beta``` folder and instead have the folder driven by a branch-specific-channel
2. Create two subfolders in every Channel folder: ```Packages``` for build consumables (nupkgs, MSIs, debian packages) and ```Intermediates``` for build intermediates (direct build outputs packaged as archives). This adds a logical separation between the build stage and the packaging stage; this gives us the freedom to rev these two pipelines independently.
3. Both ```Packages``` and ```Intermediates``` would have children folders for the specific products (Shared Framework, Host, SDK). This removes the complexity around knowing which version number means what (mainly around the '3' prefix for the version number means a different product than when the '3' is not present).
4. Each product and package folder will contain the ship train folders: Latest, LKG, Production, which will contain the relevant bits as well as a version badge and version textual file. This will allow automated systems to watch the specific train and take action when necessary, rev'ing individual products independently.

# Changes
The changes to make this happen would be contained entirely in two files: [AzurePublisher.cs](https://github.com/dotnet/cli/blob/rel/1.0.0-preview1/scripts/dotnet-cli-build/Publishing/AzurePublisher.cs) and [PublishTargets.cs](https://github.com/dotnet/cli/blob/rel/1.0.0-preview1/scripts/dotnet-cli-build/PublishTargets.cs)

# Future
Once the switch to use msbuild comes along, the only changes that should be necessary would be to the code to publish the bits to the blob locations; no external or breaking changes should be necessary.

Thoughts / comments / feedback @piotrpMSFT, @Sridhar-MS, @livarcocc, @brthor, @krwq?

### PR DESCRIPTION
**Pull request by eerhardt**
Partly fixes #3120 by organizing the output files in blob storage according to the design in the bug

@brthor @sokket

/cc @livarcocc @piotrPMSFT

### CODE REVIEW

**Comment by eerhardt** [comment]
test RHEL7.2 x64 Release Build

**Comment by livarcocc** [comment]
LGTM

File: build_projects/dotnet-cli-build/PrepareTargets.cs [hunk 1]
```diff
@@ -175,14 +175,14 @@ public static BuildTargetResult DownloadHostAndSharedFxArchives(BuildTargetConte
 
             Mkdirp(Path.GetDirectoryName(combinedSharedHostAndFrameworkArchiveDownloadFile));
 
-            if ( ! File.Exists(combinedSharedHostAndFrameworkArchiveDownloadFile))
+            if (!File.Exists(combinedSharedHostAndFrameworkArchiveDownloadFile))
```
Comment (brthor) [inline build_projects/dotnet-cli-build/PrepareTargets.cs]: taking away my style =P

Comment (livarcocc) [inline build_projects/dotnet-cli-build/PrepareTargets.cs]: It needed to go. Thanks @eerhardt!

Comment (eerhardt) [inline build_projects/dotnet-cli-build/PrepareTargets.cs]: I didn't do anything.  VS just knows what to do.

File: build_projects/dotnet-cli-build/PublishTargets.cs [hunk 2]
```diff
@@ -47,9 +48,9 @@ public static BuildTargetResult FinalizeBuild(BuildTargetContext c)
         {
             if (CheckIfAllBuildsHavePublished())
             {
-                string targetContainer = $"{Channel}/Binaries/Latest/";
+                string targetContainer = $"{AzurePublisher.Product.Sdk}/{Channel}/";
                 string targetVersionFile = $"{targetContainer}{CliNuGetVersion}";
-                string semaphoreBlob = $"{Channel}/Binaries/publishSemaphore";
+                string semaphoreBlob = $"{targetContainer}/publishSemaphore";
                 AzurePublisherTool.CreateBlobIfNotExists(semaphoreBlob);
                 string leaseId = AzurePublisherTool.AcquireLeaseOnBlob(semaphoreBlob);
```
Comment (brthor) [inline build_projects/dotnet-cli-build/PublishTargets.cs]: We need to catch here right? If this fails we can ignore it.

Comment (eerhardt) [inline build_projects/dotnet-cli-build/PublishTargets.cs]: We weren't doing it before...

I can add it, if you think that is the best approach.

File: build_projects/dotnet-cli-build/PrepareTargets.cs (hunk 1 above)
Comment (brthor) [inline build_projects/dotnet-cli-build/PrepareTargets.cs]: So many times I've skimmed over the `!` I just prefer this so it stands out. 

---
In reply to: [67968817](https://github.com/dotnet/cli/pull/3669#discussion_r67968817) [](ancestors = 67968817)

File: build_projects/dotnet-cli-build/PublishTargets.cs (hunk 2 above)
Comment (brthor) [inline build_projects/dotnet-cli-build/PublishTargets.cs]: We weren't doing it and we never saw a break until 2am on the evening of getting the release builds out. It's a rare situation but I think it's the right approach.

---
In reply to: [67969144](https://github.com/dotnet/cli/pull/3669#discussion_r67969144) [](ancestors = 67969144)

Comment (brthor) [inline build_projects/dotnet-cli-build/PublishTargets.cs]: Alternatively we can do the same as in core-setup and just wait.

---
In reply to: [67969277](https://github.com/dotnet/cli/pull/3669#discussion_r67969277) [](ancestors = 67969277,67969144)

File: build_projects/dotnet-cli-build/PrepareTargets.cs (hunk 1 above)
Comment (eerhardt) [inline build_projects/dotnet-cli-build/PrepareTargets.cs]: Get VS to change its defaults, if you think this is the best formatting.  :)

**Comment by eerhardt** [comment]
In case you are curious to the dotnet-install changes, you can see them here:

https://github.com/dotnet/cli/commit/b9ddb76249e98f9763f756a1ff732ae58001df35

They can't go in with this PR because we need at least 1 build published in the new Azure Publish format or else stage0 will be broken.  I'll open a PR for this when this PR goes in.

**Comment by eerhardt** [comment]
@brthor - I ported https://github.com/dotnet/core-setup/pull/168 in commit 2.

**Comment by brthor** [comment]
LGTM :shipit:

**Pull request merged**